docs/proposals/token-exchange-middleware.md

```diff
@@ -0,0 +1,163 @@
+
+Once the back end API token is acquired, the token is either injected into the Authorization: Bearer header or a custom header.
+
+In the first PR, we'll run the token exchange for each request. This does not scale and needs to be addressed in subsequent patches.
+
+## Usage examples
+
+```shell
+thv proxy my-mcp-server \
+  --oidc-issuer https://keycloak.example.com/realms/myrealm \
+  --oidc-client-id proxy-client \
+  --oidc-client-secret proxy-secret \
+  --token-exchange-url https://keycloak.example.com/realms/myrealm/protocol/openid-connect/token \
+  --token-exchange-client-id exchange-client \
+  --token-exchange-client-secret exchange-secret \
+  --token-exchange-audience backend-service
+```
+
+```shell
+thv run my-mcp-server \
+  --oidc-issuer https://keycloak.example.com/realms/myrealm \
+  --oidc-client-id mcp-client \
+  --token-exchange-url https://keycloak.example.com/realms/myrealm/protocol/openid-connect/token \
+  --token-exchange-client-id exchange-client \
+  --token-exchange-client-secret exchange-secret \
+  --token-exchange-audience upstream-api
+```
+
+## Operator integration
+
+For Kubernetes deployments, token exchange configuration is exposed through the `MCPServer` CRD via the `externalAuthConfig` field.
+
+### CRD structure
+
+```go
+type ExternalAuthConfig struct {
+    Type          string                 `json:"type"`  // "tokenExchange" for now
+    TokenExchange *TokenExchangeConfig   `json:"tokenExchange,omitempty"`
+}
+
+type TokenExchangeConfig struct {
+    Type      string                         `json:"type"`  // "inline" or "configMap"
+    Inline    *InlineTokenExchangeConfig     `json:"inline,omitempty"`
+    ConfigMap *ConfigMapTokenExchangeRef     `json:"configMap,omitempty"`
+}
+
+type InlineTokenExchangeConfig struct {
+    TokenURL                  string          `json:"tokenUrl"`
+    ClientID                  string          `json:"clientId"`
+    ClientSecretRef           *SecretKeyRef   `json:"clientSecretRef,omitempty"`
+    Audience                  string          `json:"audience,omitempty"`
+    Scopes                    string          `json:"scopes,omitempty"`
+    ExternalTokenHeaderName   string          `json:"externalTokenHeaderName,omitempty"`
+}
+```
+
+### Example
+
+```yaml
+apiVersion: toolhive.stacklok.dev/v1alpha1
+kind: MCPServer
+metadata:
+  name: api-proxy
+spec:
+  image: ghcr.io/my-org/mcp-server:latest
+
+  oidcConfig:
+    type: kubernetes
+    kubernetes:
+      audience: toolhive
+
+  externalAuthConfig:
+    type: tokenExchange
+    tokenExchange:
+      type: inline
+      inline:
+        tokenUrl: https://keycloak.example.com/realms/myrealm/protocol/openid-connect/token
+        clientId: exchange-client
+        clientSecretRef:
+          name: token-exchange-creds
+          key: client-secret
+        audience: backend-service
```

Comment:
I wonder if at this point we should have a dedicated CRD for the authentication pieces. I know that's out of scope of this PR, but I think we should think about it sooner rather than later. I can see folks wanting to share these configurations.

Reply:
yeah, the CRD is getting quite big. So the MCPServer CRD would have a reference to some `MCPServerAuthConfig` CRD that would be shared across servers?

We'd have to template the client names though

Reply:
I actually wonder if this /should/ be a pre-requisite though. Once we add something to our CRDs it;s hard to back off

Reply:
I'm going to merge this BUT at the same time I'm hacking on a branch to add the split external auth CRD. I'll update the design if that turns out to not be too much code (meaning not take too much time)